crates/wiggle/src/lib.rs

```diff
@@ -513,88 +513,104 @@ impl<'a, T> GuestPtr<'a, [T]> {
+    /// borrow checking is the same, but unlike `as_slice_mut`, the returned
+    /// `&mut` slice can point to WebAssembly shared memory. Though the Wiggle
+    /// borrow checker can guarantee no other Wiggle calls will access this
+    /// slice, it cannot guarantee that another thread is not modifying the
+    /// `&mut` slice in some other way. Thus, access to that slice is marked
+    /// `unsafe`.
+    pub fn as_unsafe_slice_mut(&self) -> Result<UnsafeGuestSlice<'a, T>, GuestError>
+    where
+        T: GuestTypeTransparent<'a>,
+    {
+        // Validate the bounds of the region in the original memory.
+        let len = self.checked_byte_len()?;
         let ptr =
             self.mem
                 .validate_size_align(self.pointer.0, T::guest_align(), len)? as *mut T;
-
-        let borrow = self.mem.shared_borrow(Region {
+        let region = Region {
             start: self.pointer.0,
             len,
-        })?;
+        };
 
-        // Validate all elements in slice.
-        // SAFETY: ptr has been validated by self.mem.validate_size_align
+        // Validate all elements in slice. `T::validate` is expected to be a
+        // noop for `GuestTypeTransparent` so this check may not be entirely
+        // necessary (TODO).
+        //
+        // SAFETY: `ptr` has been validated by `self.mem.validate_size_align`.
         for offs in 0..self.pointer.1 {
             T::validate(unsafe { ptr.add(offs as usize) })?;
         }
 
-        // SAFETY: iff there are no overlapping mut borrows it is valid to construct a &[T]
-        let ptr = unsafe { slice::from_raw_parts(ptr, self.pointer.1 as usize) };
-
-        Ok(GuestSlice {
+        Ok(UnsafeGuestSlice {
             ptr,
+            len: self.pointer.1 as usize,
+            region,
             mem: self.mem,
-            borrow,
         })
     }
 
-    /// Attempts to create a [`GuestSliceMut<'_, T>`] from this pointer, performing
-    /// bounds checks and type validation. The `GuestSliceMut` is a smart pointer
-    /// that can be used as a `&[T]` or a `&mut [T]` via the `Deref` and `DerefMut`
-    /// traits. The region of memory backing the slice will be marked as borrowed
-    /// by the [`GuestMemory`] until the `GuestSlice` is dropped.
+    /// Copies the data in the guest region into a [`Vec`].
     ///
-    /// This function will return a `GuestSliceMut` into host memory if all checks
-    /// succeed (valid utf-8, valid pointers, memory is not borrowed, etc). If
-    /// any checks fail then `GuestError` will be returned.
-    pub fn as_slice_mut(&self) -> Result<GuestSliceMut<'a, T>, GuestError>
+    /// This is useful when one cannot use [`GuestPtr::as_slice`], e.g., when
+    /// pointing to a region of WebAssembly shared memory.
+    ///
+    /// TODO: this copies element by element; should it be a `ptr::copy`
+    /// instead?
+    pub fn to_vec(&self) -> Result<Vec<T>, GuestError>
```

Comment:
Could this method be implemented with `as_unsafe_slice_mut` to deduplicate the bits internally?